File: backend/src/main/java/com/example/backend/pathfinding/Graph.js
class Graph {
    constructor(isDirected = false, isWeighted = false) {
        this.vertices = [];
        this.isDirected = isDirected;
        this.isWeighted = isWeighted;
    }

    addVertex(data) {
        const newVertex = new Vertex(data);
        this.vertices.push(newVertex);
        return newVertex;
    }

    removeVertex(vertex) {
        this.vertices = this.vertices.filter(v => v !== vertex);
    }

    addEdge(vertexOne, vertexTwo, weight) {
        const edgeWeight = this.isWeighted ? weight : null;

        if (vertexOne instanceof Vertex && vertexTwo instanceof Vertex) {
            vertexOne.addEdge(vertexTwo, edgeWeight);
            if (!this.isDirected) {
                vertexTwo.addEdge(vertexOne, edgeWeight);
            }
        } else {
            throw new Error('Expected Vertex arguments.');
        }
    }

    removeEdge(vertexOne, vertexTwo) {
        if (vertexOne instanceof Vertex && vertexTwo instanceof Vertex) {
            vertexOne.removeEdge(vertexTwo);
            if (!this.isDirected) {
                vertexTwo.removeEdge(vertexOne);
            }
        } else {
            throw new Error('Expected Vertex arguments.');
        }
    }

    getVertexByValue(value) {
        return this.vertices.find(vertex => vertex.data === value);
    }

    print() {
        const vertexList = this.vertices;
        vertexList.forEach(vertex => vertex.print());
    }
}

module.exports = {
    Graph,
    Vertex,
    Edge,
};

File: backend/src/main/java/com/example/backend/pathfinding/Graph.java
package com.example.backend.pathfinding;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.util.List;
import java.util.Map;

@Data
@AllArgsConstructor
@NoArgsConstructor
public class Graph {
    private Map<String,Vertex> vertices;
    private List<Edge> edges;
    public void addVertex(String vertexName){
        vertices.put(vertexName,new Vertex(vertexName,0));
    }

    public void addDirectedEdge(Edge edge){
        edges.add(edge);
    }
}
